# Introduction
This Workshop is created to give an introduction into the building of Convolutional Neural Network with the help of Tensorflow. \
Please check the Requirements in the Readme, before proceeding forward.
## Goal
The Goal is to provide a step by step solution to build a model, which seperates good an bad parts on a end-of-line model

![Alt Text](pictures/foto_Modell.png)

## steps
The workshop will include multiple steps to create a model
- inspecting the trainingsdata
- load the data and label it
- build a simple model
- improve the model
- test the model

# Imports
To work create a model and work with it a few imports are necessary

In [4]:
import os
#helpful ml modules
import matplotlib.pyplot as plt
import numpy as np
import cv2
#Tensorflow and keras
import tensorflow as tf
from tensorflow import keras
#create sequential models
from tensorflow.keras.models import Sequential
#import the layers
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, InputLayer, Dropout
#datagenerator
from tensorflow.keras.preprocessing.image import ImageDataGenerator
#augmentation
from tensorflow.keras.layers.experimental import preprocessing

### 1) inspect the datasets
There are two trainingsets in this project, one with 4 subfolders and one with 6 subfolders.
- Which assumptions can you make for the model architecture with this information?
- Load the shape of an image using the cv2 module
- which information can you take away from the shape?

In [3]:
#load the image image = cv2...
image = 
#print the image shape
print(image.shape)

(240, 320, 3)


### 2) creating datasets
To use the given datasets to train a network they need to be loaded. You can do this using cv2 or the keras ImageDataGenrator. Create one Trainings and dataset using both solutions. Decide if you want 4 or 6 classes.

#### 2a) create a dataset using cv2 and for loops
- how many lists do you need?
- how can you get to the images?
- What functions do you need from cv2
- how are you going to label your data? (hint: look at the files or folders)
- how will you normalize the data?
- print the number of images and labels

In [8]:
# trainingsset
# Initialize empty lists for data and labels

# Define the path to your training dataset directory
training_directory = 

# Iterate over subdirectories (classes) in the training dataset directory
for class_name in os.listdir(training_directory):
    class_dir = os.path.join(training_directory, class_name)

    # Iterate over image files in the class directory
    for image_name in os.listdir(class_dir):
        image_path = os.path.join(class_dir, image_name)

        # Load image

        # Normalize pixel values

        # Append image to data list


        # Assign numerical label based on filename.  Here is a if statement needed


# Print the loaded data and labels
print("Number of training samples:", len(x_train))
print("Number of training labels:", len(y_train))

Number of training samples: 840
Number of training labels: 840


expected:\
Number of training samples: 840\
Number of training labels: 840

In [10]:
# testset
# Initialize empty lists for data and labels

# Define the path to your training dataset directory
training_directory = 

# Iterate over subdirectories (classes) in the training dataset directory
for class_name in os.listdir(training_directory):
    class_dir = os.path.join(training_directory, class_name)

    # Iterate over image files in the class directory
    for image_name in os.listdir(class_dir):
        image_path = os.path.join(class_dir, image_name)

        # Load image

        # Normalize pixel values

        # Append image to data list

        # Assign numerical label based on filename. Here is a if statement needed


# Print the loaded data and labels
print("Number of test samples:", len(x_test))
print("Number of test labels:", len(y_test))

Number of test samples: 360
Number of test labels: 360


expected:\
Number of training samples: 360\
Number of training labels: 360\

#### 2b) create a dataset using ImageDataGenerator
- What functions do you need from ImageDatagenerator
- how are you going to label your data? (hint: look at the files or folders)
- how will you normalize the data?
- which configurations can be enabled?
- print the class indices

In [12]:
#trainingset
# traindirectory
data_dir_train = ''

# generate Datengeneratoren
datagen = ImageDataGenerator()

# Training Datengenerator
train_generator = datagen.

class_indices = train_generator.class_indices

# Print the class indices
print("Class indices:", class_indices)

Found 840 images belonging to 6 classes.
Class indices: {'blue': 0, 'blue_fail': 1, 'red': 2, 'red_fail': 3, 'white': 4, 'white_fail': 5}


expected:\
Found 840 images belonging to 4 classes.
Class indices: {'blue': 0, 'fail': 1, 'red': 2, 'white': 3}\
OR\
Found 840 images belonging to 6 classes.
Class indices: {'blue': 0, 'blue_fail': 1, 'red': 2, 'red_fail': 3, 'white': 4, 'white_fail': 5}


In [14]:
# Testset
# testdirectory
data_dir_eval = ''

# configure evaluationgenerator
eval_datagen = ImageDataGenerator()

# create a evaluationset
eval_generator = eval_datagen.
)
class_indices = eval_generator.class_indices

# Print the class indices
print("Class indices:", class_indices)

Found 360 images belonging to 6 classes.
Class indices: {'blue': 0, 'blue_fail': 1, 'red': 2, 'red_fail': 3, 'white': 4, 'white_fail': 5}


expected:\
Found 360 images belonging to 4 classes.
Class indices: {'blue': 0, 'fail': 1, 'red': 2, 'white': 3}\
OR\
Found 360 images belonging to 6 classes.
Class indices: {'blue': 0, 'blue_fail': 1, 'red': 2, 'red_fail': 3, 'white': 4, 'white_fail': 5}


### 3) Create a Classification model
Start with a classification model. This should be able to predict between the Classes you chose (4 or 6). Start to utilize the imported layers of Keras.
- What is the Input dimension?
- Where do you configure the Input?
- How is the last Layer configured?
- Which metric and lossfuntion do you choose?
- where is the flatten layer needed?
- start with the sequential API

In [ ]:
#start to create a Basic sequential model
model = Sequential()
#start adding layers here

#compile your model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='', metrics=[''])

### 3a) train the model
- option1: use the two lists generated with the cv2 workflow (think about the batch size)
- option2: use the datagen sets.

In [ ]:
model.fit('''put your traingsset here''', epochs=10)

### 3b)evaluate the model
Evaluate your model, again there are two options.
- How is the accuracy?
- How can it be optimized?

In [ ]:
model.evaluate('''put your traingsset here''')

### 4) Choose optimization methods if needed 
Underfitting (your model cant get a good trainings accuracy)?:

- Increase the amount of epochs
- Change the Architecture (More Dense Layers, more Conv2d Layers...)

### 
Overfitting (your model has a good training accuracy, bad a bad validation accuracy)
- Research Dropoutlayers and add them to your model
- research Keras Callback and early stopping
- Research dataaugmentation and add it to either your model or in the dataset (ImageDataGenerator needs to be reconfigured)

### 5) Export the model to .tflite
if your model has achieved a good validation accuracy you can export it to a tflite model. This can then be loaded to the controller of the conveyor belt.

In [ ]:
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

# Speichere das tflite-Modell
with open('model.tflite', 'wb') as f:
    f.write(tflite_model)

### 6) load the model on the Controller
- open the RoboCoding Pro App
- choose the option "PROJEKT IMPORTIEREN"
- import the projekt "sortingline_workshop.ft" in the folder "Fischertechnik_Projekt"
- now there should be an overview with the projekt 
- connect the Controller to the Computer with USB, for help view the file "RoboProCodingTXT40Controller.pdf" in th folder "Fischertechnik_documentation"
- press this button in the top right corner: \
![Alt Text](pictures/Verbindung.png)
- check the ip apperaing by choosing connect via usb\
![Alt Text](pictures/IP.png)
- enter the IP in the Browser
- login to the now opened webserver
- Username: ft Password: fischertechnik
- copy your tflite model to the folder custom/
- change line 19 in the process.py file of the RoboCoding project to guide to your model on the webserver
- load the programm on the Controller, by pressing the button left to the connect button
- start the programm with the play button
- error messages will be shown in the cconsol of the RoboCoding Pro App


### 7) A new Approach
After checking the first model it is time for a new approach. \
The goal is to create a model that validates the color of a piece and a probability score if it is a good part.\
The following steps need to be redone:
1. A new dataset needs to be created, use the cv2 method a think how it needs to be adapted. (Maybe a third list?)
2. create a new model architecture, this time its a functional model, not a sequential.
###
Help will be given in sequential.ipynb

#### 7a) Datasets


In [ ]:
#new dataset with cv2 and loops think about the changes.

# trainingsset
# Initialize empty lists for data and labels

# Define the path to your training dataset directory
training_directory = 

# Iterate over subdirectories (classes) in the training dataset directory
for class_name in os.listdir(training_directory):
    class_dir = os.path.join(training_directory, class_name)

    # Iterate over image files in the class directory
    for image_name in os.listdir(class_dir):
        image_path = os.path.join(class_dir, image_name)

        # Load image

        # Normalize pixel values

        # Append image to data list


        # Assign numerical label based on filename.  Here is a if statement needed


# Print the loaded data and labels
print("Number of training samples:", len(x_train))
print("Number of training labels:", len(y_train))

In [ ]:
#new dataset with cv2 and loops think about the changes.

# testset
# Initialize empty lists for data and labels

# Define the path to your training dataset directory
training_directory = 

# Iterate over subdirectories (classes) in the training dataset directory
for class_name in os.listdir(training_directory):
    class_dir = os.path.join(training_directory, class_name)

    # Iterate over image files in the class directory
    for image_name in os.listdir(class_dir):
        image_path = os.path.join(class_dir, image_name)

        # Load image

        # Normalize pixel values

        # Append image to data list

        # Assign numerical label based on filename. Here is a if statement needed


# Print the loaded data and labels
print("Number of test samples:", len(x_test))
print("Number of test labels:", len(y_test))

#### 7b functional model building

In [ ]:
# Building the model

# Convolutional Layer parameters
num_filters = 100
filter_size = (3,3)
stride = 1

# Pooling parameter
pool_size = (2,2)
# Density Layer parameters
num_dense_units = 100

# Additional parameters
brightnessfactor = 0,25
dropout_rate = 0,1
shape = (240, 320, 3)
num_classes = 3

# Define input layer
input_layer = Input(shape=shape)
x = input_layer

# Augmentation Layer


# Convolutional layers


# Output layers

# Define the model
model = Model(inputs=input_layer, outputs=[class_output, score_output])

model.summary()



#### compile youre model

In [ ]:
# Compile your model
# Extra parameter because there are 2 outputs with separate losses
loss_weights = {'class_output': 0.175, 'score_output': 0.825}

model.compile(optimizer='adam',
                      loss= {
                            'class_output': tf.keras.losses.CategoricalCrossentropy(),
                            'score_output': tf.keras.losses.BinaryCrossentropy()
                            },
                      metrics=['accuracy'],
                      loss_weights = loss_weights)

#### train
remember the datset has changed, so the parameters in the fit function shoul be adapted.

In [ ]:
model.fit()

#### evaluate and optimize

In [ ]:
model.evaluate()

#### export to tflite

In [ ]:
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

# Speichere das tflite-Modell
with open('model.tflite', 'wb') as f:
    f.write(tflite_model)

#### finally repeat number 6 and test it in action